lang/wasm_instance_cache.hh

```diff
@@ -0,0 +1,326 @@
+#include <seastar/core/timer.hh>
+#include "utils/hash.hh"
+#include "utils/overloaded_functor.hh"
+#include <chrono>
+#include <cstdint>
+#include <queue>
+#include <unordered_map>
+
+extern logging::logger wasm_logger;
+
+namespace wasm {
+
+struct wasm_instance {
+    wasmtime::Store store;
+    wasmtime::Instance instance;
+    wasmtime::Func func;
+    wasmtime::Memory memory;
+};
+
+// For each UDF full name and a scheduling group, we store a future that is ready
+// when no other task uses the same UDF instance. When an instance is retrieved
+// from the cache, the future is replaced with a new one that becomes ready when
+// the instance is returned. To avoid conflicts between tasks, each time an 
+// instance is requested, the future associated with it is removed from the cache
+// and a new one is added. When the current instance user is done with it, the
+// future that was associated with it is becomes ready.
+// This way, an instance stays in the cache as long as it isn't used by any task
+// and if there are tasks that want to use it, they get it immediately after
+// the previous user is finished.
+class instance_cache {
+public:
+    struct stats {
+        uint64_t cache_hits = 0;
+        uint64_t cache_misses = 0;
+        uint64_t cache_blocks = 0;
+    };
+
+private:
+    stats _stats;
+    seastar::metrics::metric_groups _metrics;
+
+public:
+    stats& shard_stats() {
+        return _stats;
+    }
+
+    void setup_metrics() {
+        namespace sm = seastar::metrics;
+        _metrics.add_group("user_functions", {
+            sm::make_counter("cache_hits", wasm::instance_cache::shard_stats().cache_hits,
+                            sm::description("The number of user defined function cache hits")),
+            sm::make_counter("cache_misses", wasm::instance_cache::shard_stats().cache_misses,
+                            sm::description("The number of user defined functions loaded")),
+            sm::make_counter("cache_blocks", wasm::instance_cache::shard_stats().cache_blocks,
+                            sm::description("The number of times a user defined function waited for an instance")),
+            sm::make_gauge("cache_total_size", [this] { return _total_size; },
+                            sm::description("The total size of instances stored in the user defined function cache")),
+        });
+    }
+private:
+    // key with a copied function name to store in the cache
+    struct cache_key_type {
+        seastar::scheduling_group scheduling_group;
+        cql3::functions::function_name name;
+    public:
+        cache_key_type(seastar::scheduling_group scheduling_group, const cql3::functions::function_name& name)
+            : scheduling_group(std::move(scheduling_group))
+            , name(name) {
+        }
+        bool operator==(const cache_key_type& other) const {
+            return scheduling_group == other.scheduling_group && name == other.name;
+        }
```

Comment:
Simple things like that can be done using structs. Also operator==() = default ought to work.

Reply:
(well it's a struct, but it doesn't need a constructor)

Reply:
I remember adding the default `operator==` but it didn't end up in the patch, I'll add it again. Also the constructor can in fact be omitted.